azure/durable_functions/models/NewTask.py

```diff
@@ -0,0 +1,314 @@
+            action_wrapper = WhenAnyAction
+        super().__init__(task, action_wrapper)
+
+    def try_set_value(self, child: TaskBase):
+        """Transition a WhenAny Task to a terminal state and set its value.
+
+        Parameters
+        ----------
+        child : TaskBase
+            A sub-task that just completed
+        """
+        if child.state is TaskState.SUCCEEDED:
+            if self.state is TaskState.RUNNING:
+                # the first completing sub-task sets the value
+                self.set_value(is_error=False, value=self.value)
+        else:  # child.state is TaskState.FAILED:
+            if self._first_error is None:
+                # the first failed task sets the value
+                self._first_error = child.value
+
+            # do not error out until all pending tasks have completed
+            if len(self.pending_tasks) == 0:
+                self.set_value(is_error=True, value=self._first_error)
+
+
+class RetryAbleTask(WhenAllTask):
+    """A Task representing `with_retry` scenarios.
+
+    It inherits from WhenAllTask because retryable scenarios are Tasks
+    with equivalent to WhenAll Tasks with dynamically increasing lists
+    of children. At every failure, we add a Timer child and a Task child
+    to the list of pending tasks.
+    """
+
+    def __init__(self, child: TaskBase, retry_options: RetryOptions, context):
+        self.id_ = str(child.id) + "_retryable_proxy"
+        tasks = [child]
+        super().__init__(tasks, context._replay_schema)
+
+        self.retry_options = retry_options
+        self.num_attempts = 1
+        self.context = context
+        self.actions = child.action_repr
+
+    def try_set_value(self, child: TaskBase):
+        """Transition a Retryable Task to a terminal state and set its value.
+
+        Parameters
+        ----------
+        child : TaskBase
+            A sub-task that just completed
+        """
+        if child.state is TaskState.SUCCEEDED:
+            if len(self.pending_tasks) == 0:
+                # if all pending tasks have completed,
+                # and we have a successful child, then
+                # we can set the Task's event
+                self.set_value(is_error=False, value=child.value)
+
+        else:  # child.state is TaskState.FAILED:
+            if self.num_attempts >= self.retry_options.max_number_of_attempts:
+                # we have reached the maximum number of attempts, set error
+                self.set_value(is_error=True, value=child.value)
+            else:
+                # still have some retries left.
+                # increase size of pending tasks by adding a timer task
+                # and then re-scheduling the current task after that
+                timer_task = self.context._produce_anonymous_task(parent=self)
+                self.pending_tasks.add(timer_task)
+                rescheduled_task = self.context._produce_anonymous_task(parent=self)
```

Comment:
Hmm, what do you mean by "fire immediately"? These "anonymous" tasks don't really schedule anything by themselves because, in the OOProc protocol, an API with the `WithRetry` suffix is represented as a single action object :) . So these anonymous tasks don't really add to the output payload that gets sent to the extension, they just serve for internal book-keeping and to _pretend_ as if were decomposing the `WithRetry` API into smaller tasks. We need to do this _pretending_ because our algorithm is now more close to the C# implementation which, only in this case, forces us to  keep track of redundant objects.

I know that's a lot to take in. Please let me know if this doesn't make sense! Happy to explain further and to change things up

Reply:
I see, it creates the pending tasks, which then get processed via replay as if it was a Task.WhenAll(), which makes sense. I guess the main question I have remaining is when are the corresponding Actions generated/scheduled with this approach?